### ISSUE TITLE
TypedDict accepts item access using keys of type `str`, violating PEP 589

### ISSUE THREAD

**Issue by codethief** [microsoft/pyright#10740]
**Describe the bug**
Item access `my_dict[my_key]` for TypedDict's does not lead to errors when `my_key` is of type `str`, violating the guarantees of [PEP 589](https://peps.python.org/pep-0589/#supported-and-unsupported-operations) and the [Python docs](https://typing.python.org/en/latest/spec/typeddict.html#supported-and-unsupported-operations):

> Type checkers should reject the following operations on TypedDict objects as unsafe, even though they are valid for normal dictionaries:
>
> - Operations with arbitrary `str` keys (instead of string literals or other expressions with known string values) should generally be rejected. This involves both destructive operations such as setting an item and read-only operations such as subscription expressions. As an exception to the above rule, `d.get(e)` and `e in d` should be allowed for TypedDict objects, for an arbitrary expression `e` with type `str`. The motivation is that these are safe and can be useful for introspecting TypedDict objects. The static type of `d.get(e)` should be `object` if the string value of e cannot be determined statically.

**Code or Screenshots**

```python
from typing import TypedDict

class MyDict(TypedDict):
    foo: int
    bar: str

some_dict: MyDict = {
    "foo": 3,
    "bar": "hello",
}

def get_some_unknown_str() -> str:
    raise NotImplementedError

my_key = get_some_unknown_str()
# Side remark:
# my_key: str = "non-existent"  # <-- my_key will be of type Literal["non-existent"], instead of type str?!

print(some_dict[my_key])  # <-- No type error?
print(some_dict["non-existent2"])  # <-- Type error as expected
```

Pyright only reports one type error, even though it should report two:

```
test.py:20:7 - error: Could not access item in TypedDict
    "non-existent2" is not a defined key in "MyDict" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Mypy correctly identifies both type errors:
``` 
test.py:19: error: TypedDict key must be a string literal; expected one of ("foo", "bar")  [literal-required]
test.py:20: error: TypedDict "MyDict" has no key "non-existent2"  [typeddict-item]
``` 

**VS Code extension or command-line**
I'm running `pyright[nodejs]==1.1.401` on the command line and Pylance 2025.6.2 in VSCode.

**Comment by erictraut** [comment]
Pyright's current behavior is intended and is based on significant feedback from pyright users. There are many common use cases in Python that involve the use of dynamic strings for accessing TypedDicts.

There's admittedly a tradeoff here between false positives and false negatives. In this case, we've opted to enable these use cases at the expense of potential false negatives.

**Comment by codethief** [comment]
Thanks for the quick response, @erictraut! 

> Pyright's current behavior is intended and is based on significant feedback from pyright users. There are many common use cases in Python that involve the use of dynamic strings for accessing TypedDicts.

I assume you are referring to cases such as 

```python
for key in my_dict.keys():  # key is of type `str`
    print(my_dict[key])
```

?

It looks like I'm the minority, but I'm disappointed Pyright is deviating from the PEP here. 

If item access using `str` keys really is desired, I would argue this would much rather be a case for `# type: ignore`, for using `my_dict.get(key)`, or for additional type narrowing (of the key) or type widening (of the dictionary), i.e. changes on the user's side, than having Pyright deviate from the official contract. Or for making [TypedDict even stricter](https://peps.python.org/pep-0728/#the-closed-class-parameter) and thus improving type inference for `.keys()`.

The TypeScript community has already played through pretty much this exact situation and to this day bugs due to `noUncheckedIndexedAccess: false` are quite common in my experience. At the same time, it's anything but easy to migrate an existing code base from `noUncheckedIndexedAccess: false` to `true` once people have opened Pandora's box. 

However, in Pyright it seems I don't even have the option to strictly follow the PEP behavior(?) (`typeCheckingMode: "strict"` does cause Pyright to report a type error in my above code snippet but at least it is not the one I would expect.)
